mpa/det/inferrer.py

```diff
@@ -159,30 +159,15 @@ def dump_features_hook(mod, inp, out):
         def dummy_dump_features_hook(mod, inp, out):
             feature_vectors.append(None)
 
-        def dump_saliency_hook(model: torch.nn.Module, input: Tuple, out: List[torch.Tensor]):
-            """ Dump the last feature map to `saliency_maps` cache
-
-            Args:
-                model (torch.nn.Module): PyTorch model
-                input (Tuple): input
-                out (List[torch.Tensor]): a list of feature maps
-            """
-            with torch.no_grad():
-                saliency_map = get_saliency_map(out[-1])
-            saliency_maps.append(saliency_map.squeeze(0).detach().cpu().numpy())
-
-        def dummy_dump_saliency_hook(model, input, out):
-            saliency_maps.append(None)
-
         feature_vector_hook = dump_features_hook if dump_features else dummy_dump_features_hook
-        saliency_map_hook = dump_saliency_hook if dump_saliency_map else dummy_dump_saliency_hook
 
         # Use a single gpu for testing. Set in both mm_val_dataloader and eval_model
         if is_module_wrapper(model):
             model = model.module
         with eval_model.module.backbone.register_forward_hook(feature_vector_hook):
-            with eval_model.module.backbone.register_forward_hook(saliency_map_hook):
+            with SaliencyMapHookDet(eval_model.module) if dump_saliency_map else nullcontext() as shook:
                 eval_predictions = single_gpu_test(eval_model, data_loader)
+                saliency_maps = shook.records if dump_saliency_map else [None] * len(self.dataset)
```

Comment:
How about using `saliency_hook` instead of `shook`?

Reply:
Updated!